### ISSUE TITLE
Persist Across Reboots?

### ISSUE THREAD

**Issue by disruptivepatternmaterial** [dsclee1/BRmesh-esp32-mqtt#1]
Hello---

Is there a way to have this persist across reboots of the esp32? It seems like every time I power cycle it, I have to re join the lights?

If this is a known issue / WAD, cool :) otherwise let me know what debug info I can share.

Thanks!

**Comment by dsclee1** [comment]
Glad someone's giving this a try! 

Unfortunately that's a known issue at the moment, should have added it to the list of bugs. 

Need to add code that writes back the stored lights to the ESP32.

**Comment by disruptivepatternmaterial** [comment]
yeah I am tinkered with persisting the info, but not got it working...I think it needs to be on the file system given the amount of data...nit I could be wrong.

**Comment by Shaztech** [comment]
Check my issue that I closed. You only need to keep the same "my_key" don't use a random one.

**Comment by dsclee1** [comment]
Good point @Shaztech , and good to see you sorted your other issue.

Think that might also require a few changes to the auto-discovery code to ensure the existing lights are rediscovered and available via MQTT though?

**Comment by Shaztech** [comment]
I don't use MQTT so I don't know. On my end I only have 2 lights that are controlled at the same time (same color, state... etc) so I only send the command and all the lights react at the same time.

**Comment by codeclinic** [comment]
Having to readjust everything in HA etc every times the esop32 boots is a bit of a hassle. If you could work this out that'd be great.

Not knowing much about how this code works, would it not be possible to set a flag after the first discovery. Then if that flag is present, it skips discovery and uses the existing keys etc. If new devices need to be added it'd then be a simple case of just uploading a clean image to the esp32 via Vis studio?

**Comment by disruptivepatternmaterial** [comment]
Yeah this turns out to be not-so-simple, I thew mine away and just got Zigbee lights that are the same from GLEDOPTO. It is a shame because the bulk of this code is really complex and CLEARLY a lot of work went into it.

But you would need to maybe persist the light configs somewhere - on the esp32 or on HA/MQTT somewhere that can be fetched on startup.

Just reusing the key vs generating a new one on each cycle is useful but it is about 10% of the solution...

**Comment by ericcastro** [comment]
I have managed to persist the paired lights across reboots of the ESP32 using the internal storage using the `Preferences` built in library.

Tinkering with this until 3:30 AM allowed me to get a pretty good understanding of everything the code does and how the BRMesh protocol works, as well as creating virtual MQTT devices for HA, so I am confident I will be able to share this soon with everyone who's been waiting.

I need to polish the code, and get it tested enough. So far worked only on 1 single light, I will try persisting my entire 6 light setup..

I need to get some sleep now :P

**Comment by codeclinic** [comment]
> I am confident I will be able to share this soon with everyone

Please keep us updated. I look forward to trying this out. Let me know if you need help testing on a multi light setup.

Thanks!

**Comment by codeclinic** [comment]
> I am confident I will be able to share this soon with everyone

Please keep us updated. I look forward to trying this out. Let me know if you need help testing on a multi light setup.

Thanks!

**Comment by ericcastro** [comment]
By the way I'm taking a bit longer as I haven't had much time and also noticed the original code does not handle warm/cold light. You can notice because only regular colors work, and no matter how close you get you can never match the original warm/cold brightness as in the BRMesh app - what this tells me is that the lamps have separate internal leds for warm/cold that are brighter but are not yet supported in the ESP32 implementation.

I will have to tinker a bit more to see if I can add support for that too

**Comment by codeclinic** [comment]
Thanks for keeping us updated. Let me know if there is anything I can do to help.

**Comment by GitHub-Noub** [comment]
Hi @ericcastro , were you able to commit the changes to maintain persistency/manage warm/cold light? Nearing Christmas so dusted off the old lights and came across @dsclee1 's excellent work in getting this off an app and into MQTT/HA which works really well (once I'd understood how to get it working) but having the persistency/warm/cold colour control would be the icing on the floodlight cake.

**Comment by ericcastro** [comment]
Hi, I have a working version that persists and supports newer types of bulbs that were not detected by the original code. I have had it working super stable for months now, and resists resets, power outages etc

I'm so sorry I completely forgot about this thread.

There is only one caveat, which I haven't been able to solve: I can only pair 2 lightbulbs. Once I get the 3rd one, the ESP32 drops the WiFi connection.

The code changed dramatically from the original, I will be looking to polish it with AI now and publish.

If anyone is interested please let me know on here, I will gladly send you a zip file and/or link to a repo as soon as I have it up.
